velox/type/Variant.cpp

```diff
@@ -194,9 +198,9 @@ std::string variant::toJson(const TypePtr& type) const {
           b += ",";
         }
         b += "{\"key\":";
-        b += pair.first.toJson();
+        b += pair.first.toJson(type->childAt(0));
         b += ",\"value\":";
-        b += pair.second.toJson();
+        b += pair.second.toJson(type->childAt(1));
```

Comment:
FYI - This breaks koski , Since it allows us to create variants like this : `{"key1": {"key2": [null, 2], "key3": (3, 4), "key4": "five"}}` 
The type of value vector need not be same in the map. 

cc: @mbasmanova @majetideepak 

I propose changing this to `pair.second.toJson(pair.second.type())` .

Reply:
Velox Map types are strongly typed where the value vector is the same.
Can you share how Velox consumes or generates the variant mentioned?

Reply:
Not sure I fully understand your question. Variant as it is written now allows you to have maps like the one I posted , See: `{"key1": {"key2": [null, 2], "key3": (3, 4), "key4": "five"}}` . 

This is used internally in koski (a python dataframe api which uses velox::variant) - which allows you to create dataframes for the above kind of data. Currently this change breaks this functionality , which might affect large number of use cases. 

One way to resolve this is to have say toJsonUnsafe() which is the old implemention that is called by pyToJson. 
I am open to other ideas here.

Reply:
@kgpai thanks! It looks like Velox::variant maps can be untyped.
We can probably modify as `std::string toJson(const std::optional<TypePtr> type)`.
The comment must say `toJson()` without a `TypePtr` will infer the types recursively.
Inside `std::string toJson(const std::optional<TypePtr> type)`, we check if the type is present, else we use `inferType()`.
Further, before we recurse, we pass the child type if `type` is present or we pass `nullopt`.
`pyToJson()` will invoke `toJson(nullopt)`.
Since you already imported this code, can you please make this change and merge it? Thanks.

Reply:
Thanks @majetideepak . One of my concerns with this is that making the type optional will encourage use where the type wont be passed . Maybe this can be controlled inside velox, but it will be harder to police inside Meta. Secondly despite comments, the original reason why this change was made might become tribal knowledge. Having an explicit function that is marked _unsafe_ and is only used for this one use case makes it easier to maintain and hopefully get rid of sooner sooner. So in my opinion though ugly I think its better to have the old implementation living beside this but explicitly marked as unsafe (say toJsonUnsafe()) with comments on why it exists etc. 

@mbasmanova Let me know what your opinion is.

Reply:
@kgpai Krishna, I agree with your reasoning. toJsonUnsafe() seems a better option to me.

Reply:
@kgpai sounds good!